README.md
# FlipDrawer

An easy way to use drawer in Flutter with cool flipping animation.

![](example.gif)

## Getting Started

Add dependency in your flutter project.

```
$ flutter pub add flip_drawer
```

or

```yaml
dependencies:
  flip_drawer: ^1.0.3
```

Then run `flutter pub get`.

## Example Project

You can see the `example` folder to learn many different ways of using FlipDrawer in your app.

## Usage

As usual, we need to import the package.

```dart
import 'package:flip_drawer/flip_drawer.dart';
```

Then wrap the app home page with FlipDrawer. 

To use the basic FlipDrawer, you need to define title and items (List of MenuItem) to generate menu in your drawer. The drawer will use your default theme color as drawer background, and theme brightness as drawer brightness.

```dart
class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: title,
      // FlipDrawer will use style from this theme
      theme: ThemeData(
        brightness: Brightness.dark,
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      // wrap your HomePage with FlipDrawer
      home: FlipDrawer(
        items: [
          MenuItem('Home', onTap: (){}),
          MenuItem('Project', onTap: (){}),
          MenuItem('Favourite', onTap: (){}),
          MenuItem('Profile', onTap: (){}),
          MenuItem('Setting', onTap: (){}),
        ],
        title: 'Your App Title',
        child: HomePage(),
      ),
    );
  }
}
```

Then set AppBar to only use a bare AppBar without title and leading. Title and menu button will be managed by FlipDrawer.

```dart
appBar: AppBar(),
```

### Icons

We can also use icon for menu items.

```dart
home: FlipDrawer(
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  title: 'Your App Title',
  child: HomePage(),
),
```

### Custom Head Drawer

It's advisable to set alignment to FlipDrawerAlignment.start when you use headDrawer.

```dart
home: FlipDrawer(
  title: 'Your App Title',
  alignment: FlipDrawerAlignment.start,
  headDrawer: Image.asset('boys.png'),
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  child: HomePage(),
),
```

### Custom Content Drawer

In the previous example, we use items and let the FlipDrawer generate menu in the content drawer.
But we can also use any custom widget as the content drawer.

```dart
home: FlipDrawer(
  contentDrawer: Container(
    padding: EdgeInsets.symmetric(horizontal: 5),
    child: Column(
      children: [
        ListTile(title: 'Home', icon: Icons.home, onTap: (){}),
        ListTile(title: 'Project', icon:Icons.rss_feed, onTap: (){}),
        ListTile(title: 'Favourite', icon: Icons.favorite_border, onTap: (){}),
        ListTile(title: 'Profile', icon: Icons.person_outline, onTap: (){}),
        ListTile(title: 'Setting', icon: Icons.settings, onTap: (){}),
      ],
    ),
  ),
  title: 'Your App Title',
  child: HomePage(),
),
```

### Custom Full Drawer

If you have your own creative idea for the full widget inside the drawer, don't worry. We can use it in FlipDrawer by passing the full widget to the drawer. But make sure to use the same background color for the custom drawer as for the Theme color.

```dart
home: FlipDrawer(
  drawer: Container(
    color: Colors.teal,
    padding: EdgeInsets.symmetric(vertical: 36, horizontal: 15),
    child: Theme(
      data: ThemeData(brightness: Brightness.dark),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.end,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          ListTile(title: 'Home', icon: Icons.home, onTap: (){}),
          ListTile(title: 'Project', icon:Icons.rss_feed, onTap: (){}),
          ListTile(title: 'Favourite', icon: Icons.favorite_border, onTap: (){}),
          ListTile(title: 'Profile', icon: Icons.person_outline, onTap: (){}),
          ListTile(title: 'Setting', icon: Icons.settings, onTap: (){}),
        ],
      ),
    ),
  ),
  title: 'Your App Title',
  child: HomePage(),
),
```

### Others

We can also use custom curve and duration. The default values are linear and 250ms. We can specify curveReverse and durationReverse to change curve and duration of animation in the reverse direction.

```dart
home: FlipDrawer(
  curve: Curves.easeInOut,
  duration: Duration(milliseconds: 200),
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  title: 'Your App Title',
  child: HomePage(),
),
```

There it is. Have fun!!!

CHANGELOG.md
## [1.0.3] - 22 June 2021

* auto close drawer when tap on menu item

## [1.0.2] - 19 May 2021

* Migrate to null safety

## [1.0.1] - 18 Sept 2020

* Upgrade curved_animation_controller to the new version

## [1.0.0] - 18 Sept 2020

* A working FlipDrawer


lib/src/drawer.dart
import 'dart:math';

import 'package:curved_animation_controller/curved_animation_controller.dart';
import 'package:flutter/material.dart';
import 'package:flip_drawer/src/alignment.dart';
import 'package:flip_drawer/src/container.dart';
import 'package:flip_drawer/src/item.dart';

class FlipDrawer extends StatefulWidget {
  /// The color to use for the drawer background. Typically this should be set
  /// along with [brightness].
  ///
  /// If this is null, then [ThemeData.primaryColor] is used.
  //final Color backgroundColor;

  /// The brightness of the drawer. Typically this is set along
  /// with [backgroundColor], [backgroundGradient].
  ///
  /// If this is null, then [ThemeData.primaryColorBrightness] is used.
  //final Brightness brightness;

  /// This is where you should attach your main application widget
  final Widget child;

  /// Custom drawer to be used when you don't want to use
  /// the default [SlideDrawerContainer] generated from [items] or [contentDrawer]
  final Widget? drawer;

  /// Head drawer to be rendered before [contentDrawer]
  /// or the default generated content drawer from [items]
  final Widget? headDrawer;

  /// Content drawer to be used if you don't want to use
  /// the default content drawer generated from [items]
  final Widget? contentDrawer;

  /// List of [MenuItem] to be used to generate the default content drawer
  final List<MenuItem> items;

  /// Duration of the drawer sliding animation
  ///
  /// Default is [Duration(milliseconds: 300)]
  final Duration duration;

  /// Curve to be used for the drawer sliding animation
  final Curve curve;

  /// Duration of the drawer sliding animation in the reverse direction
  final Duration? reverseDuration;

  /// Curve to be used for the drawer sliding animation in the reverse direction
  final Curve? reverseCurve;

  /// Vertical alignment of content inside drawer
  /// it can [start] from the top, or [center]
  final FlipDrawerAlignment? alignment;

  /// Offset from right to calculate the end point of sliding animation
  ///
  /// Default is 60.0
  final double offsetFromRight;

  final String title;

  final bool isAnimateMenuButton;

  const FlipDrawer({
    Key? key,
    this.items = const [],
    this.drawer,
    this.headDrawer,
    this.contentDrawer,
    required this.child,
    this.duration = const Duration(milliseconds: 250),
    this.curve = Curves.easeInOut,
    this.reverseDuration,
    this.reverseCurve,
    this.alignment,
    required this.title,
    this.offsetFromRight = 60.0,
    this.isAnimateMenuButton = true,
  }) : super(key: key);

  static _FlipDrawerState? of(BuildContext context) =>
      context.findAncestorStateOfType<_FlipDrawerState>();

  @override
  _FlipDrawerState createState() => _FlipDrawerState();
}

class _FlipDrawerState extends State<FlipDrawer> with TickerProviderStateMixin {
  bool _canBeDragged = false;
  double get _maxSlide =>
      MediaQuery.of(context).size.width - widget.offsetFromRight;

  bool get _hasReverseCurve => widget.reverseCurve != null;
  bool get _hasReverseDuration => widget.reverseDuration != null;

  late CurvedAnimationController _animation;
  late CurvedAnimationController _menuAnimation;

  ThemeData get theme => Theme.of(context);

  @override
  void initState() {
    super.initState();
    _initAnimation();
  }

  @override
  void dispose() {
    _animation.dispose();
    _menuAnimation.dispose();
    super.dispose();
  }

  _initAnimation() {
    _animation = CurvedAnimationController(
      vsync: this,
      duration: widget.duration,
      curve: widget.curve,
      reverseCurve: _hasReverseCurve ? widget.reverseCurve : widget.curve,
      reverseDuration:
          _hasReverseDuration ? widget.reverseDuration : widget.duration,
      debugLabel: 'FlipDrawer',
    );

    _menuAnimation = CurvedAnimationController(
      vsync: this,
      duration: widget.duration,
      curve: widget.curve,
      reverseCurve: _hasReverseCurve ? widget.reverseCurve : widget.curve,
      reverseDuration:
          _hasReverseDuration ? widget.reverseDuration : widget.duration,
    );

    _animation.addListener(() => setState(() {}));
    _menuAnimation.addListener(() => setState(() {}));
  }

  open() {
    _animation.start();
    if (widget.isAnimateMenuButton) _menuAnimation.start();
  }

  close() {
    _animation.reverse();
    if (widget.isAnimateMenuButton) _menuAnimation.reverse();
  }

  toggle() => _animation.isCompleted ? close() : open();

  _onDragStart(DragStartDetails details) {
    bool isDragOpenFromLeft = _animation.isDismissed;
    bool isDragCloseFromRight = _animation.isCompleted;

    _canBeDragged = isDragOpenFromLeft || isDragCloseFromRight;
  }

  _onDragUpdate(DragUpdateDetails details) {
    if (_canBeDragged && details.primaryDelta != null) {
      double delta = details.primaryDelta! / _maxSlide;
      _animation.progress += delta;
      if (widget.isAnimateMenuButton) _menuAnimation.progress += delta;
    }
  }

  _onDragEnd(DragEndDetails details) {
    double _kMinFlingVelocity = 365.0;

    if (_animation.isDismissed || _animation.isCompleted) {
      return;
    }

    if (details.velocity.pixelsPerSecond.dx.abs() >= _kMinFlingVelocity) {
      double visualVelocity = details.velocity.pixelsPerSecond.dx /
          MediaQuery.of(context).size.width;
      _animation.fling(velocity: visualVelocity);
      if (widget.isAnimateMenuButton)
        _menuAnimation.fling(velocity: visualVelocity);
    } else if (_animation.progress < 0.5) {
      close();
    } else {
      open();
    }
  }

  Widget get _drawer => FlipDrawerContainer(
        alignment: widget.alignment,
        head: widget.headDrawer,
        content: widget.contentDrawer,
        drawer: widget.drawer,
        items: widget.items,
        paddingRight: widget.offsetFromRight,
        headPaddingRight: _animation.progress < 0.9
            ? widget.offsetFromRight
            : (1 - _animation.progress) * widget.offsetFromRight,
      );

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        if (_animation.isCompleted) {
          close();
          return Future.value(false);
        }
        return Future.value(true);
      },
      child: GestureDetector(
        onHorizontalDragStart: _onDragStart,
        onHorizontalDragUpdate: _onDragUpdate,
        onHorizontalDragEnd: _onDragEnd,
        child: Material(
          color: theme.primaryColor,
          child: Stack(
            children: [
              Transform(
                transform: Matrix4.identity()
                  ..translate(_maxSlide * (_animation.value - 1), 0.0),
                child: Transform(
                  transform: Matrix4.identity()
                    ..setEntry(3, 2, 0.001)
                    ..rotateY(pi / 2 * (1 - _animation.value)),
                  alignment: Alignment.centerRight,
                  child: _drawer,
                ),
              ),
              Transform(
                transform: Matrix4.identity()
                  ..translate(_maxSlide * _animation.value, 0.0),
                child: Transform(
                  transform: Matrix4.identity()
                    ..setEntry(3, 2, 0.001)
                    ..rotateY(-pi * _animation.value / 2),
                  alignment: Alignment.centerLeft,
                  child: widget.child,
                ),
              ),
              Positioned(
                top: 16.0 + MediaQuery.of(context).padding.top,
                left: _animation.value * MediaQuery.of(context).size.width,
                width: MediaQuery.of(context).size.width,
                child: Text(
                  widget.title,
                  style: Theme.of(context).primaryTextTheme.headline6,
                  textAlign: TextAlign.center,
                ),
              ),
              Positioned(
                top: 4.0 + MediaQuery.of(context).padding.top,
                left: 4.0 + _animation.value * _maxSlide,
                child: IconButton(
                  icon: AnimatedIcon(
                    icon: AnimatedIcons.menu_close,
                    color: Colors.white,
                    progress: _menuAnimation.controller,
                  ),
                  onPressed: toggle,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/src/container.dart
import 'package:flutter/material.dart';
import 'package:flip_drawer/src/alignment.dart';
import 'package:flip_drawer/src/item.dart';

import 'drawer.dart';

class FlipDrawerContainer extends StatelessWidget {
  final Widget? drawer;

  final Widget? head;

  final Widget? content;

  final List<MenuItem> items;

  final double paddingRight;

  final double headPaddingRight;

  /// The color to use for the background. Typically this should be set
  /// along with [brightness].
  ///
  /// If this property is null, then [ThemeData.primaryColor] is used.
  //final Color backgroundColor;

  /// The brightness of the app bar's material. Typically this is set along
  /// with [backgroundColor], [backgroundGradient].
  ///
  /// If this property is null, then [ThemeData.primaryColorBrightness] is used.
  //final Brightness brightness;

  /// Vertical alignment of content inside [SlideDrawerContainer]
  /// it can [start] from the top, or [center]
  final FlipDrawerAlignment? alignment;

  bool get _hasItems => items.isNotEmpty;
  bool get _hasDrawer => drawer != null;
  bool get _hasHead => head != null;
  bool get _hasContent => content != null;

  FlipDrawerContainer({
    Key? key,
    this.drawer,
    this.head,
    this.content,
    this.items = const [],
    //this.brightness,
    //this.backgroundColor,
    this.alignment,
    this.paddingRight = 0.0,
    this.headPaddingRight = 0.0,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    ThemeData theme = Theme.of(context);
    FlipDrawerAlignment? _alignment = alignment;

    if (_alignment == null) {
      _alignment =
          _hasHead ? FlipDrawerAlignment.start : FlipDrawerAlignment.center;
    }

    bool _isAlignTop = _alignment == FlipDrawerAlignment.start;

    double _width = MediaQuery.of(context).size.width;
    double _scale = (_width - headPaddingRight) / _width;

    return Material(
      child: _hasDrawer
          ? drawer
          : Container(
              decoration: BoxDecoration(color: theme.primaryColor),
              child: SafeArea(
                child: Theme(
                  data: ThemeData(brightness: theme.primaryColorBrightness),
                  child: Column(
                    mainAxisAlignment: _isAlignTop
                        ? MainAxisAlignment.start
                        : MainAxisAlignment.center,
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      if (_hasHead)
                        Transform(
                          transform: Matrix4.identity()
                            ..scale(_scale, 1.0, 1.0),
                          child: head,
                        ),
                      if (_hasContent)
                        Container(
                          margin: EdgeInsets.only(right: paddingRight),
                          child: content,
                        ),
                      if (!_hasContent && _hasItems)
                        for (MenuItem item in items)
                          Container(
                            margin: EdgeInsets.only(right: paddingRight),
                            child: MenuItemWidget(item: item),
                          ),
                    ],
                  ),
                ),
              ),
            ),
    );
  }
}

class MenuItemWidget extends StatelessWidget {
  final MenuItem item;

  MenuItemWidget({Key? key, required this.item}) : super(key: key);

  Widget? get _leading {
    if (item.hasLeading) return item.leading!;
    if (item.hasIcon) return Icon(item.icon);
    return null;
  }

  @override
  Widget build(BuildContext context) {
    return ListTile(
      leading: _leading,
      contentPadding: EdgeInsets.only(left: _leading == null ? 24 : 16),
      title: Text(item.title),
      onTap: () {
        if (item.isCloseDrawerWhenTapped) {
          FlipDrawer.of(context)?.close();
        }
        item.onTap?.call();
      },
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flip_drawer/flip_drawer.dart';

void main() {
  runApp(App());
}

enum FlipDrawerExampleType {
  BASIC,
  ICON,
  HEAD_DRAWER,
  CONTENT_DRAWER,
  HEAD_CONTENT_DRAWER,
  FULL_DRAWER,
}

class App extends StatefulWidget {
  @override
  _AppState createState() => _AppState();
}

class _AppState extends State<App> {
  Key _appKey = UniqueKey();
  FlipDrawerExampleType type = FlipDrawerExampleType.BASIC;
  String _title = 'Flip Drawer Demo';

  bool get _isBasic => type == FlipDrawerExampleType.BASIC;
  bool get _isIcon => type == FlipDrawerExampleType.ICON;
  bool get _isHead => type == FlipDrawerExampleType.HEAD_DRAWER;
  bool get _isContent => type == FlipDrawerExampleType.CONTENT_DRAWER;
  bool get _isHeadContent => type == FlipDrawerExampleType.HEAD_CONTENT_DRAWER;
  bool get _isFullDrawer => type == FlipDrawerExampleType.FULL_DRAWER;

  List<MenuItem> get _items => [
        MenuItem('Basic',
            onTap: () => _changeType(FlipDrawerExampleType.BASIC)),
        MenuItem('With Icon',
            onTap: () => _changeType(FlipDrawerExampleType.ICON)),
        MenuItem('Custom Header',
            onTap: () => _changeType(FlipDrawerExampleType.HEAD_DRAWER)),
        MenuItem('Custom Content',
            onTap: () => _changeType(FlipDrawerExampleType.CONTENT_DRAWER)),
        MenuItem('Header and Content',
            onTap: () =>
                _changeType(FlipDrawerExampleType.HEAD_CONTENT_DRAWER)),
        MenuItem('Full Drawer',
            onTap: () => _changeType(FlipDrawerExampleType.FULL_DRAWER)),
      ];

  List<MenuItem> get _itemsIcon => [
        MenuItem('Basic',
            icon: Icons.rss_feed,
            onTap: () => _changeType(FlipDrawerExampleType.BASIC)),
        MenuItem('With Icon',
            icon: Icons.favorite_border,
            onTap: () => _changeType(FlipDrawerExampleType.ICON)),
        MenuItem('Custom Header',
            icon: Icons.map,
            onTap: () => _changeType(FlipDrawerExampleType.HEAD_DRAWER)),
        MenuItem('Custom Content',
            icon: Icons.person_outline,
            onTap: () => _changeType(FlipDrawerExampleType.CONTENT_DRAWER)),
        MenuItem('Header and Content',
            icon: Icons.alarm,
            onTap: () =>
                _changeType(FlipDrawerExampleType.HEAD_CONTENT_DRAWER)),
        MenuItem('Full Drawer',
            icon: Icons.settings,
            onTap: () => _changeType(FlipDrawerExampleType.FULL_DRAWER)),
      ];

  _changeType(type) {
    if (this.type != type)
      setState(() {
        this.type = type;
        _appKey = UniqueKey();
      });
  }

  Widget get _home {
    if (_isBasic) return _basic;
    if (_isIcon) return _icon;
    if (_isHead) return _head;
    if (_isContent) return _content;
    if (_isHeadContent) return _headContent;
    if (_isFullDrawer) return _fullDrawer;
    return _basic;
  }

  Widget get _basic => FlipDrawer(
        title: _title,
        child: HomePage(),
        items: _items,
      );

  Widget get _icon => FlipDrawer(
        title: _title,
        child: HomePage(),
        items: _itemsIcon,
      );

  Widget get _head => FlipDrawer(
        title: _title,
        child: HomePage(),
        headDrawer: Image.asset('boys.png'),
        alignment: FlipDrawerAlignment.start,
        items: _itemsIcon,
      );

  Widget get _content => FlipDrawer(
        title: _title,
        child: HomePage(),
        contentDrawer: Container(
          padding: EdgeInsets.symmetric(horizontal: 5),
          child: Column(
            children: [
              for (var item in _itemsIcon)
                ListTile(
                  title: Text(item.title),
                  leading: Icon(item.icon),
                  onTap: item.onTap,
                ),
            ],
          ),
        ),
      );

  Widget get _headContent => FlipDrawer(
        title: _title,
        child: HomePage(),
        alignment: FlipDrawerAlignment.start,
        headDrawer: Image.asset('boys.png'),
        contentDrawer: Container(
          padding: EdgeInsets.only(left: 10, top: 18),
          child: Column(
            children: [
              for (var item in _itemsIcon)
                ListTile(
                  title: Text(item.title),
                  leading: Icon(item.icon),
                  onTap: item.onTap,
                ),
            ],
          ),
        ),
      );

  Widget get _fullDrawer => FlipDrawer(
        title: _title,
        child: HomePage(),
        offsetFromRight: 80,
        isAnimateMenuButton: false,
        drawer: Container(
          color: Colors.teal,
          padding: EdgeInsets.symmetric(vertical: 36, horizontal: 15),
          child: Theme(
            data: ThemeData(brightness: Brightness.dark),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.end,
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                for (var item in _items)
                  ListTile(
                    title: Text(item.title),
                    onTap: item.onTap,
                  ),
              ],
            ),
          ),
        ),
      );

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      key: _appKey,
      title: _title,
      theme: ThemeData(
        primarySwatch: Colors.teal,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: _home,
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
          // Here we take the value from the MyHomePage object that was created by
          // the App.build method, and use it to set our appbar title.
          //title: Text('Flip Drawer Demo'),
          //leading: Container(),
          ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
